# Regressão 01 - tarefa 02 - transformações em X e Y

### 1. Melhorando o ajuste no percentual de gorjetas

a. Vamos partir do modelo feito na última tarefa, o que relaciona ```tip_pct``` e ```net_bill```. Carregue a base, os pacotes e reproduza este modelo aqui.

In [1]:
import pandas as pd
import seaborn as sns
from seaborn import load_dataset
from sklearn.metrics import r2_score

import matplotlib.pyplot as plt

import numpy as np

import statsmodels.formula.api as smf

In [2]:
tips = sns.load_dataset("tips")
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']
tips.head()

total_bill   tip     sex smoker  day    time  size   tip_pct  net_bill
0       16.99  1.01  Female     No  Sun  Dinner     2  0.063204     15.98
1       10.34  1.66    Male     No  Sun  Dinner     3  0.191244      8.68
2       21.01  3.50    Male     No  Sun  Dinner     3  0.199886     17.51
3       23.68  3.31    Male     No  Sun  Dinner     2  0.162494     20.37
4       24.59  3.61  Female     No  Sun  Dinner     4  0.172069     20.98

In [3]:
# reproduzindo a regressão: tip ~ net_bill
modelo = smf.ols('tip ~ net_bill', data=tips).fit()
print(modelo.summary())

                            OLS Regression Results                            
Dep. Variable:                    tip   R-squared:                       0.333
Model:                            OLS   Adj. R-squared:                  0.330
Method:                 Least Squares   F-statistic:                     120.6
Date:                Tue, 19 Aug 2025   Prob (F-statistic):           5.02e-23
Time:                        15:02:27   Log-Likelihood:                -375.63
No. Observations:                 244   AIC:                             755.3
Df Residuals:                     242   BIC:                             762.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.3307      0.168      7.907      0.0

### 2. ajuste o modelo de ```tip_pct``` em função do logaritmo de ```net_bill```.

In [5]:
modelo_log = smf.ols('tip_pct ~ np.log(net_bill)', data=tips).fit()
print(modelo_log.summary())

                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.202
Model:                            OLS   Adj. R-squared:                  0.198
Method:                 Least Squares   F-statistic:                     61.16
Date:                Tue, 19 Aug 2025   Prob (F-statistic):           1.62e-13
Time:                        15:09:14   Log-Likelihood:                 123.81
No. Observations:                 244   AIC:                            -243.6
Df Residuals:                     242   BIC:                            -236.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.6181      0.054  

###  3. ajuste o modelo de ```tip_pct``` em função de um polinômio de segundo grau em ```net_bill```.

In [7]:
modelo_poli = smf.ols('tip_pct ~ net_bill + I(net_bill**2)', data=tips).fit()
print(modelo_poli.summary())

                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.158
Model:                            OLS   Adj. R-squared:                  0.151
Method:                 Least Squares   F-statistic:                     22.55
Date:                Tue, 19 Aug 2025   Prob (F-statistic):           1.06e-09
Time:                        15:30:08   Log-Likelihood:                 117.25
No. Observations:                 244   AIC:                            -228.5
Df Residuals:                     241   BIC:                            -218.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.4851      0.047  

###  4. ajuste um modelo no log do ```tip_pct``` em função de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```

In [8]:
tips['log_tip_pct'] = np.log(tips['tip_pct'])

modelo_log_tip = smf.ols('log_tip_pct ~ net_bill', data=tips).fit()
print(modelo_log_tip.summary())

                            OLS Regression Results                            
Dep. Variable:            log_tip_pct   R-squared:                       0.235
Model:                            OLS   Adj. R-squared:                  0.232
Method:                 Least Squares   F-statistic:                     74.48
Date:                Tue, 19 Aug 2025   Prob (F-statistic):           8.27e-16
Time:                        15:30:48   Log-Likelihood:                -106.00
No. Observations:                 244   AIC:                             216.0
Df Residuals:                     242   BIC:                             223.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.2749      0.056    -22.872      0.0

In [9]:
y_pred = np.exp(modelo_log_tip.fittedvalues)
r2 = r2_score(tips['tip_pct'], y_pred)
print("R² no tip_pct:", r2)

R² no tip_pct: 0.08676609740629226


###  5. ajuste um modelo no log do ```tip_pct``` em função do log de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```.

In [10]:
tips['log_net_bill'] = np.log(tips['net_bill'])

modelo_loglog = smf.ols('log_tip_pct ~ log_net_bill', data=tips).fit()
print(modelo_loglog.summary())

                            OLS Regression Results                            
Dep. Variable:            log_tip_pct   R-squared:                       0.293
Model:                            OLS   Adj. R-squared:                  0.291
Method:                 Least Squares   F-statistic:                     100.5
Date:                Tue, 19 Aug 2025   Prob (F-statistic):           5.20e-20
Time:                        15:31:04   Log-Likelihood:                -96.357
No. Observations:                 244   AIC:                             196.7
Df Residuals:                     242   BIC:                             203.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -0.3942      0.133     -2.961   

In [11]:
y_pred = np.exp(modelo_loglog.fittedvalues)
r2 = r2_score(tips['tip_pct'], y_pred)
print("R² no tip_pct:", r2)

R² no tip_pct: 0.2276852951813606


###  6. qual modelo resultou no melhor ajuste?

In [13]:
print("R² modelo linear (net_bill):", r2_score(tips['tip_pct'], modelo.fittedvalues))
print("R² modelo polinômio:", r2_score(tips['tip_pct'], modelo_poli.fittedvalues))
print("R² modelo log-tip:", r2)
print("R² modelo log-log:", r2)

R² modelo linear (net_bill): -320.89194090439173
R² modelo polinômio: 0.15760784734296185
R² modelo log-tip: 0.2276852951813606
R² modelo log-log: 0.2276852951813606


-> O melhor ajuste foi obtido pelos modelos com transformações logarítmicas 'log-tip' e 'log-log', que alcançaram R² = 0.23, aproximadamente. Embora ainda não seja um ajuste muito forte, eles explicam melhor a variação de 'tip_pct' em comparação aos modelos linear e polinomial.